File: 0075-sort-colors/solution.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        divides(nums,0,nums.size()-1);
        print(nums,nums.size());
    }
        void merge(vector<int> &arr, int low, int mid, int high){
        vector<int>newarr;
        int i = low;
        int j = mid+1;

        while(i<=mid && j<=high){
            if(arr[i]<arr[j]){
                newarr.push_back(arr[i]);
                i++;
            }else{
                newarr.push_back(arr[j]);
                j++;
            }
        }
        while(i<=mid){
            newarr.push_back(arr[i]);
            i++;
        }
        while(j<=high){
            newarr.push_back(arr[j]);
            j++;
        }
        for (int k = low; k<=high; k++){
            arr[k] = newarr[k-low];
        }
    }
        void divides(vector<int>&arr, int low, int high){
            if(low<high){
                int mid = (low+high)/2;
                divides(arr,low,mid);
                divides(arr,mid+1,high);
                merge(arr,low,mid,high);
            }
        }
        void print(vector<int>&arr,int size){
            for (int i = 0; i<size; i++){
                cout<<arr[i] << " ";
            }
        }
};

File: 0075-sort-colors/solution.java
class Solution {
    public void sortColors(int[] nums) {
        int left = 0;
        int right = nums.length - 1;
        int mid = 0;

        while (mid <= right) {
            if (nums[mid] == 0) {
                // Only XOR if indices are different to avoid resetting to 0
                if (mid != left) {
                    nums[mid] ^= nums[left];
                    nums[left] ^= nums[mid];
                    nums[mid] ^= nums[left];
                }
                left++;
                mid++;
            } else if (nums[mid] == 2) {
                // Only XOR if indices are different
                if (mid != right) {
                    nums[mid] ^= nums[right];
                    nums[right] ^= nums[mid];
                    nums[mid] ^= nums[right];
                }
                
                right--; 
            } else {
                mid++;
            }
        }
    }
}

